### PR TITLE
use binary search to optimize performance

### PR DESCRIPTION
**Pull request by Lan-ce-lot**
Use binary search to optimize performance in pkg/ksyms/ksyms.go

### CODE REVIEW

**Review by kkourt — CHANGES_REQUESTED** [review]
Thanks!

Since binary search can be tricky to implement in some cases (e.g., https://ai.googleblog.com/2006/06/extra-extra-read-all-about-it-nearly.html), can we please add some unit tests for this?

**Comment by Lan-ce-lot** [comment]
Ok, I will add test.

**Comment by Lan-ce-lot** [comment]
I added some tests, the previous code and the current code have passed. And I prevents overflow at the same time.
```
m := l + ((r - l + 1) >> 1)
```
PTAL, thanks! @kkourt

File: pkg/ksyms/ksyms_test.go [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+		},
+		{
+			name: "addr 0x900",
+			addr: 0x900,
+			want: FnOffset{SymName: "addr8", Offset: 0x100},
+		},
+		{
+			name:    "address is before first symbol",
+			addr:    0x100,
+			wantErr: true,
```
Comment (kkourt) [inline pkg/ksyms/ksyms_test.go:59]: Here, I believe we need to return the  symbol with an offset of zero.

**Review by kkourt — CHANGES_REQUESTED** [review]
Thanks!

I believe that the correct behavior if a user requests the address of a symbol is to return this symbol with an offsset of zero.

Could you also add a Signed-off-by with a real name? See: https://docs.cilium.io/en/v1.13/contributing/development/contributing_guide/#dev-coo

File: pkg/ksyms/ksyms_test.go (hunk 1 above)
Comment (Lan-ce-lot) [reply pkg/ksyms/ksyms_test.go:64]: Ok, I have modified.

**Comment by Lan-ce-lot** [comment]
> Thanks!
> 
> I believe that the correct behavior if a user requests the address of a symbol is to return this symbol with an offset of zero.
> 
> Could you also add a Signed-off-by with a real name? See: https://docs.cilium.io/en/v1.13/contributing/development/contributing_guide/#dev-coo

Sure!

**Comment by Lan-ce-lot** [comment]
I added a Signed-off-by with a real name.
PTAL, Thanks! @kkourt

**Review by kkourt — APPROVED** [review]
LGTM, thanks!

**Pull request merged**